### ISSUE TITLE
Segmentation mask is too small for large detected objects (1280,500) on image size (1280, 720)

### ISSUE THREAD

**Issue by SimHerling** [ultralytics/ultralytics#13554]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi, I was looking for a similar question and I looked into docs, but I could not find anything so I'm opening this issue.

Problem summary: The width of the segmentation mask is too small for large detected objects (1280,500) on image size 1280, model used gelan-c-seg.yaml

I have trained a segmentation model on video size 1280x720 (model settings are below).
The detection is run on size 1280 as well. 
Small and medium size objects are detected very well and the segmentation masks work very well.

The problem is with objects which are large (1280, 500) pixels. The width of such an object goes across the whole image (size 1280 x 720), i.e. the width of the object is 1280 pixels, its height goes usually from the top third of the image to the bottom, i.e. the height of the object is around 500 pixels. The predicted mask (and bounding box), however, do not cover the whole object. The bbox and mask should have a size (1280, 500) (w, h) but they have a size (900, 500). The problem is only with the width, the height works well.

Training settings for gelan-c-seg.yaml':
segment/train.py
--cfg 'models/segment/gelan-c-seg.yaml'
--weights 'segment/yolov9c-seg.pt'
--imgsz 1280

Segment prediction:
segment/predict.py
--imgsz 1280
-- retina_masks

Do you know why this could be? I have used in the past also the yolov7 segmentation model (settings:  --cfg yolov7-seg.yaml 
--weights yolov7-seg.pt) and it was able to detect complete masks for large objects (1280,500) on image size 1280x720 so I am wondering why it is not working with the model gelan-c-seg.yaml.

Many thanks for help

**Comment by SimHerling** [comment]
Thanks for the links and document, but I have not found there an answer for my question. Can you please advise..

**Comment by glenn-jocher** [comment]
@SimHerling hello,

Thank you for reaching out and providing detailed information about your issue. It sounds like you've done a thorough job in setting up your segmentation model and identifying the problem with large object masks.

To help you better, let's go through a few steps:

1. **Reproducible Example**: Could you please provide a minimum reproducible code example? This will help us replicate the issue on our end and investigate further. You can refer to our guide on creating a minimum reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example). This step is crucial for us to understand the exact setup and parameters you're using.

2. **Version Check**: Ensure that you are using the latest versions of `torch` and `ultralytics`. Sometimes, bugs are fixed in newer releases, and upgrading might resolve the issue. You can upgrade your packages using the following commands:
    ```bash
    pip install --upgrade torch ultralytics
    ```

3. **Model Configuration**: The issue might be related to the specific configuration of the `gelan-c-seg.yaml` model. It would be helpful to compare the configurations between the `gelan-c-seg.yaml` and the `yolov7-seg.yaml` models to identify any significant differences that might affect the segmentation mask size. Ensure that the model architecture and hyperparameters are suitable for handling large objects.

4. **Retina Masks**: Since you are using the `--retina_masks` option, ensure that this setting is correctly implemented in your model configuration. Retina masks can sometimes behave differently based on the model architecture and the size of the objects being detected.

5. **Training Data**: Verify that your training data includes a sufficient number of large objects and that the annotations are accurate. The model might not generalize well to large objects if it hasn't seen enough examples during training.

6. **Post-Processing**: Check if there are any post-processing steps that might be affecting the mask size. Sometimes, the bounding box and mask might be adjusted during post-processing, leading to discrepancies.

Here is an example of how you might set up your training and prediction commands:

```python
from ultralytics import YOLO

# Load the model
model = YOLO('segment/yolov9c-seg.pt')

# Train the model
model.train(data='path/to/your/data.yaml', imgsz=1280, epochs=50, batch=4)

# Predict using the model
results = model.predict(source='path/to/your/image.jpg', imgsz=1280, retina_masks=True)
```

Please provide the reproducible example and ensure your packages are up-to-date. This will help us assist you more effectively. If the issue persists, we can delve deeper into the model configurations and training data.

Thank you for your patience and cooperation. We're here to help you resolve this issue!

**Comment by SimHerling** [comment]
Hi Glen, thanks for the tips, I have tried a few in my code, but nothing helped.

I have therefore created this MRE (below) so that you can replicate the issue.
I have used the sample image 'horses.jpg' from the yolo model samples.

The segmentation masks work well on the original image. I have resized and cropped this image 3 times:
- img_1280_xl.jpg
- img_1280_xxl.jpg
- img_1280_xxxl.jpg'

with increasing focus on the white horse in the frontend of the image.

While the masks in the original image horses.jpg and the first resized image img_1280_xl.jpg worked well, the last 2 images img_1280_xxl.jpg and img_1280_xxxl.jpg where the white horse took up almost the whole part of the image the masks did not work well. The masks covered just part of the horse and the rest of the horse was not covered by the mask. This is exactly the same problem which I see in my own custom model.

I hope the details I provided help..
Many thanks

MRE:

!pip install --upgrade torch ultralytics
!pip install dill

from ultralytics import YOLO
from PIL import Image

model = YOLO('segment/yolov9c-seg.pt')
img = Image.open('data/images/horses.jpg')

# Resize and crop original image
w, h = 1280, 720
size_1280 = (w, h)
size_3840 = (w * 3, h * 3)

img_3840 = img.resize(size_3840)

img_xl = img_3840.crop((0, 400, 3000, 400 + 3000/(16/9)))
img_xxl = img_3840.crop((0, 750, 1800, 750 + 1800/(16/9)))
img_xxxl = img_3840.crop((0, 800, 1700, 800 + 1700/(16/9)))


img_1280_xl = img_xl.resize(size_1280)
img_1280_xxl = img_xxl.resize(size_1280)
img_1280_xxxl = img_xxxl.resize(size_1280)

img_1280_xl.show()
img_1280_xxl.show()
img_1280_xxxl.show()

img_1280_xl.save('data/images/img_1280_xl.jpg')
img_1280_xxl.save('data/images/img_1280_xxl.jpg')
img_1280_xxxl.save('data/images/img_1280_xxxl.jpg')


# Run inference on an the original image and the resized images

results = model.predict('data/images/horses.jpg', save=True, imgsz=1280, name='horses_mask_issue')
results = model.predict('data/images/img_1280_xl.jpg', save=True, imgsz=1280, name='horses_mask_issue')
results = model.predict('data/images/img_1280_xxl.jpg', save=True, imgsz=1280, name='horses_mask_issue')
results = model.predict('data/images/img_1280_xxxl.jpg', save=True, imgsz=1280, name='horses_mask_issue')

**Comment by glenn-jocher** [comment]
Hi @SimHerling,

Thank you for providing a detailed Minimum Reproducible Example (MRE) and the images to illustrate the issue. This is very helpful! 😊

### Steps to Reproduce
Based on your MRE, it looks like the segmentation masks are not scaling correctly for larger objects in resized images. Let's ensure a few things:

1. **Latest Versions**: Ensure you are using the latest versions of `torch` and `ultralytics`. You can upgrade them using:
    ```bash
    !pip install --upgrade torch ultralytics
    ```

2. **Model Configuration**: Sometimes, model configurations might need adjustments for handling larger objects. Ensure that the model configuration (`gelan-c-seg.yaml`) is optimized for your use case.

### Code Example
Here’s a slightly modified version of your code to ensure everything is set up correctly:

```python
from ultralytics import YOLO
from PIL import Image

# Ensure the latest versions are installed
!pip install --upgrade torch ultralytics

# Load the model
model = YOLO('segment/yolov9c-seg.pt')

# Open the original image
img = Image.open('data/images/horses.jpg')

# Resize and crop the original image
w, h = 1280, 720
size_1280 = (w, h)
size_3840 = (w * 3, h * 3)

img_3840 = img.resize(size_3840)

img_xl = img_3840.crop((0, 400, 3000, 400 + 3000/(16/9)))
img_xxl = img_3840.crop((0, 750, 1800, 750 + 1800/(16/9)))
img_xxxl = img_3840.crop((0, 800, 1700, 800 + 1700/(16/9)))

img_1280_xl = img_xl.resize(size_1280)
img_1280_xxl = img_xxl.resize(size_1280)
img_1280_xxxl = img_xxxl.resize(size_1280)

# Save the resized images
img_1280_xl.save('data/images/img_1280_xl.jpg')
img_1280_xxl.save('data/images/img_1280_xxl.jpg')
img_1280_xxxl.save('data/images/img_1280_xxxl.jpg')

# Run inference on the original and resized images
results = model.predict('data/images/horses.jpg', save=True, imgsz=1280, name='horses_mask_issue')
results = model.predict('data/images/img_1280_xl.jpg', save=True, imgsz=1280, name='horses_mask_issue')
results = model.predict('data/images/img_1280_xxl.jpg', save=True, imgsz=1280, name='horses_mask_issue')
results = model.predict('data/images/img_1280_xxxl.jpg', save=True, imgsz=1280, name='horses_mask_issue')
```

### Next Steps
1. **Check Model Configuration**: Ensure that the `gelan-c-seg.yaml` configuration is suitable for handling large objects. You might need to adjust parameters related to the mask size or resolution.

2. **Retina Masks**: Since you are using `--retina_masks`, ensure that this setting is correctly implemented and optimized in your model configuration.

3. **Training Data**: Verify that your training data includes a sufficient number of large objects and that the annotations are accurate. The model might not generalize well to large objects if it hasn't seen enough examples during training.

4. **Post-Processing**: Check if there are any post-processing steps that might be affecting the mask size. Sometimes, the bounding box and mask might be adjusted during post-processing, leading to discrepancies.

If the issue persists, please let us know, and we can delve deeper into the model configurations and training data. Thank you for your patience and cooperation. We're here to help you resolve this issue!

**Comment by SimHerling** [comment]
Thanks @glenn-jocher,

I have gone through the recommended steps (details below) but I am not getting good masks for large objects.

I did another simple check with 4 images of cars (attaching below) which take up almost the whole image. This is with the default simplest model config and default pre-trained segment weights (MRE below) but it seems the predicted masks are incomplete for large cars spanning across the whole image. Any idea why?

MRE - cars:

!pip install --upgrade torch ultralytics
!pip install dill

from ultralytics import YOLO
from PIL import Image

model = YOLO('segment/yolov9c-seg.pt')

# Run inference
results = model.predict(['car2.jpeg','car8.jpg','car9.jpeg', 'car10.jpeg'], save=True, imgsz=1280, name='cars_profile)

re: Step to Reproduce:
1. I did install !pip install --upgrade torch ultralytics
2. what should be changed in the config of the gelan-c-seg.yaml to predict segmentation masks for large instances? I was hoping that this config already predicts the large images

re: Next Steps
1. what exactly should I change? Even for the simplified use case of cars mentioned in the MRE - cars
2. I am not using --retina_masks in neither the MRE with horses nor with cars
3. The training data is from pretrained weights 'yolov9c-seg.pt' so it should not be a problem
4. No postprocessing is being done.


Incomplete masks for large cars on images:

![car2](https://github.com/ultralytics/ultralytics/assets/107577652/84616ade-6a25-474f-856d-7e6c3e248a6b)
![car10](https://github.com/ultralytics/ultralytics/assets/107577652/d119736d-3aef-415e-b1ab-6a40b6ced9e7)
![car9](https://github.com/ultralytics/ultralytics/assets/107577652/91b2c777-2173-438b-a08a-64ca04029394)
![car8](https://github.com/ultralytics/ultralytics/assets/107577652/276dbc54-ae99-4e31-ab89-0f3b4c514e1e)

**Comment by glenn-jocher** [comment]
Hi @SimHerling,

Thank you for providing the additional details and the Minimum Reproducible Example (MRE). This is very helpful! 😊

### Steps to Verify
1. **Latest Versions**: You've confirmed that you're using the latest versions of `torch` and `ultralytics`, which is great. This ensures we're working with the most up-to-date code and bug fixes.

2. **Model Configuration**: It appears that the default configuration might not be optimized for large objects. Let's look into potential adjustments.

### Potential Solutions
1. **Model Configuration Adjustments**: The `gelan-c-seg.yaml` configuration might need some tweaks to handle large objects better. Here are a few suggestions:
    - **Anchor Sizes**: Ensure that the anchor sizes in your model configuration are suitable for large objects. You might need to adjust the anchor sizes to better fit the dimensions of large objects.
    - **Image Size**: While you're using `imgsz=1280`, ensure that the model's input size during training and inference is consistent. Larger input sizes can help in better segmenting large objects.

2. **Training Data**: Although you're using pre-trained weights, the model might benefit from fine-tuning on a dataset that includes large objects. This can help the model learn to segment large objects more accurately.

### Example Configuration Adjustment
Here's an example of how you might adjust the anchor sizes in your `gelan-c-seg.yaml` configuration:

```yaml
anchors:
  - [10, 13, 16, 30, 33, 23]  # Adjust these values based on your object sizes
  - [30, 61, 62, 45, 59, 119]
  - [116, 90, 156, 198, 373, 326]
```

### Code Example
Here's a modified version of your MRE to include the configuration adjustments:

```python
from ultralytics import YOLO
from PIL import Image

# Ensure the latest versions are installed
!pip install --upgrade torch ultralytics

# Load the model
model = YOLO('segment/yolov9c-seg.pt')

# Run inference on the car images
results = model.predict(['car2.jpeg', 'car8.jpg', 'car9.jpeg', 'car10.jpeg'], save=True, imgsz=1280, name='cars_profile')
```

### Next Steps
1. **Anchor Sizes**: Adjust the anchor sizes in your model configuration to better fit large objects.
2. **Fine-Tuning**: Consider fine-tuning the model on a dataset that includes large objects to improve segmentation accuracy.

If the issue persists, please let us know, and we can delve deeper into the model configurations and training data. Thank you for your patience and cooperation. We're here to help you resolve this issue!

**Comment by SimHerling** [comment]
Thanks @glenn-jocher for the tips for the next steps.

I have replaced in the file gelan-c-seg.yaml the default anchors = 3 with the below anchors, but I do not see any effect on the large objects (the small and medium work very well with the new anchors):

anchors:
  - [10, 13, 16, 30, 100, 50]
  - [30, 61, 62, 45, 200, 70]
  - [116, 90, 156, 198, 1200, 400]

My data set contains small objects, medium and large objects, these are their typical sizes:

small objects - width range in pixels (4, 20), height range (4,20), the ratio w:h is approximately 1:1
medium objects - width range in pixels (100, 500), height (50,100), the ratio w:h is approximately 2:1 to 5:1
large objects - width range (1200, 1280), height (400, 600), the ratio w:h is approximately 2.5:1

I found some some info in the forums about anchors but no good complete instructions, would you have a recommendation how to change the above 3 lines of code with anchor values to get better predictions for my large objects? I understand this may need fine tuning and its hard to determine them well, but knowing the typical sizes I listed down some good starting values may be great.

Also I saw discussions about reg_max parameter. Do you think it is relevant for my model to change it? I have not found details in which .py files the reg_max parameter shall be changed, if you thing it may help then any instructions where to change it would be great. Many thanks.

**Comment by glenn-jocher** [comment]
Hi @SimHerling,

Thank you for the detailed follow-up and for providing the additional context. It's great to hear that you've made some progress with the image size adjustment! 😊

### Anchor Adjustments
Given your dataset's object size distribution, fine-tuning the anchors is indeed crucial. Your current anchor settings are a good start, but let's refine them further based on the typical sizes you've provided. Here’s a suggested configuration:

```yaml
anchors:
  - [10, 13, 16, 30, 50, 75]  # Small objects
  - [30, 61, 62, 45, 100, 150]  # Medium objects
  - [116, 90, 156, 198, 300, 400]  # Large objects
```

### Fine-Tuning Anchors
Fine-tuning anchors can be iterative. Here’s a step-by-step approach:
1. **Start with the suggested anchors**: Use the above configuration as a starting point.
2. **Train and Evaluate**: Train your model and evaluate the performance on your validation set.
3. **Adjust Anchors**: Based on the performance, adjust the anchors incrementally. For example, if large objects are still not detected well, increase the anchor sizes slightly.

### reg_max Parameter
The `reg_max` parameter is used in some advanced YOLO models to control the maximum regression value for bounding box coordinates. Adjusting this can help with bounding box precision, especially for large objects. However, it’s not always necessary to tweak this unless you’re seeing significant issues with bounding box accuracy.

If you decide to experiment with `reg_max`, you can typically find and adjust it in the model configuration files or the training script. Here’s a general approach:
1. **Locate the Parameter**: Look for `reg_max` in your model configuration file or the training script.
2. **Adjust the Value**: Start with small adjustments and observe the impact on your model’s performance.

### Example Code for Training with Adjusted Anchors
Here’s an example of how you might set up your training script with the adjusted anchors:

```python
from ultralytics import YOLO

# Load the model
model = YOLO('segment/yolov9c-seg.pt')

# Adjust the anchors in the model configuration
model.model.yaml['anchors'] = [
    [10, 13, 16, 30, 50, 75],
    [30, 61, 62, 45, 100, 150],
    [116, 90, 156, 198, 300, 400]
]

# Train the model
model.train(data='path/to/your/data.yaml', imgsz=1280, epochs=50, batch=4)
```

### Next Steps
1. **Experiment with Anchors**: Use the suggested anchors and adjust based on performance.
2. **Evaluate reg_max**: If necessary, experiment with the `reg_max` parameter for better bounding box precision.

Feel free to share your results or any further questions you might have. We're here to help you optimize your model for the best performance! 🚀
